### DISCUSSION TITLE
Create an simple JS code in Node with RxJS

### DISCUSSION THREAD

**Discussion by GregorBiswanger** [swissmanu/rxjs-debugging-for-vscode#125] [General]
Hi there,

I think your idea is really great! Thanks for the great work.

I created a new simple node project without webpack etc. and unfortunately it doesn't work with plain js. Then I tried TypeScript and got your breakpoints offered. Unfortunately, no information is displayed here when executing via debugging. It would be interesting for me to experiment and teach for training courses.

How could you get it going?

![image](https://user-images.githubusercontent.com/7336300/144509967-f9f686b8-5718-43f4-a171-2fe535d6d28b.png)

**Comment by swissmanu** [comment]
Hi @GregorBiswanger 

Thank you for giving RxJS Debugging a try!

> I created a new simple node project without webpack etc. and unfortunately it doesn't work with plain js. Then I tried TypeScript and got your breakpoints offered.

As of now, the extension requires TypeScript, that's correct. Though thinking of it right now, there might be an easy way how plain JavaScript code could be supported as well. I opened an issue for this: #126 

> Unfortunately, no information is displayed here when executing via debugging.

Can you provide more information about your development environment, please?

- Operating System
- Visual Studio Code Version
- "RxJS Debugging for Visual Studio Code" Extension Version
- NodeJS Version

> It would be interesting for me to experiment and teach for training courses.

Oh, cool! 🥰 This was one intention of RxJS Debugging: Lower the entry boundary for RxJS newcomers.

**Comment by GregorBiswanger** [comment]
Operating System: **Windows 10**
Visual Studio Code Version: **1.62.3**
"RxJS Debugging for Visual Studio Code" Extension Version: **v1.0.1**
NodeJS Version: **16.13.0**

I just created a new package.json and installed **rxjs 7.4.0** and **typescript 4.5.2** (via pnpm).

The tsconfig.json looks like this:

```
{
    "compilerOptions": {
        "module": "commonjs",
        "sourceMap": true,
        "lib": ["ES2021", "DOM"]
    },
    "exclude": [
        "node_modules"
    ]
}
```
transpiled and started a nodejs debugging with source map support via vscode. 

```
    "version": "0.2.0",
    "configurations": [
        {
            "type": "pwa-node",
            "request": "launch",
            "name": "Launch Program",
            "skipFiles": [
                "<node_internals>/**"
            ],
            "sourceMaps": true,
            "program": "${workspaceFolder}\\sample.js",
            "outFiles": [
                "${workspaceFolder}/**/*.js"
            ]
        }
    ]
```

Unfortunately no information from the extension.

**Reply by swissmanu** [reply]
Thank you for the details 👍
Looks all good so far ☹️ Is there a way how you could provide me your playground via a Git repository? I'd love to have a look myself.

Regarding "Work with Plain JavaScript": I just released version [1.1.0](https://github.com/swissmanu/rxjs-debugging-for-vscode/blob/main/CHANGELOG.md) which works without TypeScript files. 🙌

**Comment by GregorBiswanger** [comment]
I also tried version 1.1.0. Unfortunately does not work. I haven't tried plain js yet.

Here I have published my project on GitHub:
[https://github.com/GregorBiswanger/RxJSDebuggingExample](https://github.com/GregorBiswanger/RxJSDebuggingExample)

A matching video:
[https://github.com/GregorBiswanger/RxJSDebuggingExample/raw/main/example-debugging.mp4](https://github.com/GregorBiswanger/RxJSDebuggingExample/raw/main/example-debugging.mp4)

Then I have another question. I am the producer of [www.rxjs-fruits.com](https://www.rxjs-fruits.com). Here I use Monaco Editor (VSCode) - If you have an idea how we would get your extension in here, it would certainly be helpful when learning RxJS. What do you think?

**Reply by swissmanu** [reply]
> Then I have another question. I am the producer of www.rxjs-fruits.com. Here I use Monaco Editor (VSCode) - If you have an idea how we would get your extension in here, it would certainly be helpful when learning RxJS. What do you think?

I love RxJS-fruits! We used it at @Altoo-AG as preparation material for our junior frontend engineers, which had no RxJS experience at all. Their feedback was positive all the way.

Now to your question: Though some components of RxJS Debugging (e.g., the runtimes and the whole CDP communication part) were built to be reusable independently from vscode, I don't see an easy way of porting the extension "as a whole" to plain Monaco.

But maybe this would go to far anyway? What if you'd just use operator log points as an inspiration and implement a similar functionality specifically for RxJS-fruit with Monaco instead? Many things described in [ARCHITECTURE.md](https://github.com/swissmanu/rxjs-debugging-for-vscode/blob/main/ARCHITECTURE.md) add a lot of complexity which should not be necessary for your case.

Let me know if you are interested in discussing this idea further. We could set up a call or something.

**Reply by GregorBiswanger** [reply]
Yes, unfortunately Monaco does not have the extensions host. I think it won't be that easy. I will definitely take a look at your documentation. Thanks for your great work! Even if it doesn't work for me yet. Would use that very well for my trainings and lectures on RXJS. Greetings from Nuremberg :)

**Reply by GregorBiswanger** [reply]
It works! 🥳🥳🥳
It goes very quickly of course.. It would be nice if you could step back and forth like a recording.. ala Time Travel Debugging (See the Quokka extension)..

Now I would like to test whether it works with Angular, for example ... do you have to take something extra into account?!

![image](https://user-images.githubusercontent.com/7336300/145280056-3093eba0-9f68-4f70-89b1-49681a7dd2bf.png)

**Reply by GregorBiswanger** [reply]
Automatically generating marble diagrams would also be awesome...

**Comment by GregorBiswanger** [comment]
Did you test my project?!
https://github.com/GregorBiswanger/RxJSDebuggingExample

I would like to use your extension... :)

**Reply by swissmanu** [reply]
Good morning. Yes, I just had a look; It took me a bit longer to find time since I am switching projects/context over the week :-)
Thank you for making your code available.

I saw two problems so far:

1. There seems to be a problem with resolve the source maps, which I don't fully understand yet. I created a bug for this: #136 
2. The enabled log points are not synced to the runtime early enough: `of` emits "too soon". Once I changed the source observable to something async like `interval` things started to work again. This is something that needs to be fixed: #137

Thank you for your valuable feedback ❤️ And sorry for the disappointing experience on your side.

**Comment by GregorBiswanger** [comment]
I have now tried your example from README in JS. Unfortunately, it doesn't work either.

![image](https://user-images.githubusercontent.com/7336300/145204489-1c711f8f-c899-4a04-9563-b9c164ad9cd5.png)

**Reply by swissmanu** [reply]
This one should work once v1.1.1 is released.

A very stupid mistake, but a quickfix: https://github.com/swissmanu/rxjs-debugging-for-vscode/commit/1cdb1579b872243a10747c94d9c623759dfa83f0#diff-c67d2ed8e133f2645f3ae15f05c51638829156d9c04317e39a0f6de73b57413e

**Reply by GregorBiswanger** [reply]
cool!  Can't wait to try it out at home... 🚀🚀🚀

**Comment by GregorBiswanger** [comment]
Now I also have TypeScript without SourceMaps. Unfortunately it doesn't work either.

![image](https://user-images.githubusercontent.com/7336300/145204822-87038795-e7fa-4689-9740-b00b5d0463ce.png)
